python_modules/dagster/dagster/_core/definitions/assets/definition/asset_effect.py

```diff
@@ -67,9 +67,10 @@ class AssetEffect(Effect[AssetSpec]):
     @abstractmethod
     def execution_type(self) -> AssetExecutionType: ...
 
-    def to_out(self, can_subset: bool) -> Out:
+    def to_out(self, can_subset: bool, annotation_type: Any) -> Out:
         return Out(
-            dagster_type=self.spec.metadata.get(SYSTEM_METADATA_KEY_DAGSTER_TYPE),
+            dagster_type=self.spec.metadata.get(SYSTEM_METADATA_KEY_DAGSTER_TYPE)
```

Comment:
continuing [1], we'd prob check that `self.spec.metadata.get(SYSTEM_METADATA_KEY_DAGSTER_TYPE)` is a more constrained type than `annotation_type`, if both are non-`None`